KAFKA-13873 Add ability to Pause / Resume KafkaStreams Topologies

jnh5y opened the pull request:
This PR adds the ability to pause and resume KafkaStreams instances as well as named/modular topologies.

Added an integration test to show how pausing and resuming works.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jnh5y committed:
KAFKA-13873 Add ability to Pause / Resume KafkaStreams Topologies

This PR adds the ability to pause and resume KafkaStreams instances as well as named/modular topologies.

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -463,7 +463,10 @@ public void restore(final Map<TaskId, Task> tasks) {
                     if (restoreChangelog(changelogs.get(partition))) {
-                        tasks.get(taskId).clearTaskTimeout();
+                        final Task task = tasks.get(taskId);
+                        if (task != null) {
```
Why are we getting null here when we were not before?

jnh5y replied:
The taskId is coming from the changeLogs partition info a few lines above.  The tasks map returned is now smaller and does not contain all the keys anymore!

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -897,7 +897,8 @@ private void initializeAndRestorePhase() {
         // if there's no active restoring or standby updating it would not try to fetch any data
-        changelogReader.restore(taskManager.tasks());
+        // After KAFKA-13873, we only restore the not paused tasks.
+        changelogReader.restore(taskManager.notPausedTasks());
```
when a paused task is resumed will restore be called if necessary?

jnh5y replied:
"Yes."

The next time that a resumed task goes through the loop it will be returned in `notPausedTasks()` and `restore` will be called.  (I say "Yes" because that's not a separate, direct call to `restore` when resuming.)

jnh5y replied:
I think I've covered this with a new test in `StreamThreadTest`.

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:
```diff
@@ -273,6 +273,12 @@ Collection<Task> allTasks() {
     }
 
+    Collection<Task> notPausedTasks() {
+        return Collections.unmodifiableCollection(readOnlyActiveTasks.stream().filter(t ->
```
nit: Can we align each call on its own line for readability?

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,897 @@
+    private static Properties producerConfig;
+    private static Properties consumerConfig;
+
+    private static final Materialized<Object, Long, KeyValueStore<Bytes, byte[]>> IN_MEMORY_STORE = Materialized.as(Stores.inMemoryKeyValueStore("store"));
```
nit: line length

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+        assertEquals(0, (long) maxLag);
+
+        // Verify no output somehow? (Hard to prove negative)
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 0), CoreMatchers.equalTo(Collections.emptyList()));
```
Maybe you can wait the max poll interval?

jnh5y replied:
After some work, we've found a slightly better way to work through this.

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+
+        // Create KafkaStream
+        final StreamsBuilder builder = new StreamsBuilder();
+        //builder.stream(EXISTING_STREAM).groupBy((k, v) -> k).count(IN_MEMORY_STORE).toStream().to(OUTPUT_STREAM_1);
```
nit: clean up?

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 0), CoreMatchers.equalTo(Collections.emptyList()));
+
+        // Pause KafkaStreams/topplogy
+        kafkaStreams.pause();
```
do we need to pause twice?

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+        CLUSTER.createTopic(NEW_STREAM, 2, 1);
+
+        // Create KafkaStream / NamedTopologyBuilders
+        final KafkaStreamsNamedTopologyWrapper streams = new KafkaStreamsNamedTopologyWrapper(props());
```
I don't see where this gets cleaned up...

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+    @After
+    public void shutdown() {
+        if (kafkaStreams != null) {
+            kafkaStreams.close(Duration.ofSeconds(30));
```
you can probably just call close()

jnh5y replied:
I'll admit that I cribbed from other tests.  It looks like calling `close()` calls this method with `close(Long.MAX_VALUE)`.  I think this is fine as-is.

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+        // Use statelistener to verify transition to paused transition?
+        // Read status as Paused?
+        // JNH: Assert the topology1 is paused and topology2 is not paused
+        assertTrue(streams.isNamedTopologyPaused("topology1"));
```
Make sure the other topology is not paused?

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+        // Verify that topology 2 is paused and has processed data.
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_2, 0), CoreMatchers.equalTo(Collections.emptyList()));
+
+        // Verify that topology 2 is not paused and has processed data.
```
clean up

wcarlson5 commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,501 @@
+        assertTrue(streams.isNamedTopologyPaused("topology2"));  // Still paused
+
+        // Verify that the input written during the pause is processed.
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 3), CoreMatchers.equalTo(COUNT_OUTPUT_DATA2));
```
topology2 as well

wcarlson5 reviewed (commented):
looking good for the most part just had a couple questions. 

I gave the tests a quick pass. I realize you are still working on them so feel free to ignore that part until you are ready

jnh5y commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,440 @@
+        kafkaStreams = new KafkaStreams(builder.build(props()), props());
+
+        // Start KafkaStream with paused processing.
+        kafkaStreams.pause();
+        kafkaStreams.start();
```
I think the names here are a little confusing.  With context, it makes sense that this is how to start a `KafkaStreams` instance with processing paused.

If anyone has a naming suggestion here, I'm very open to it!

bvarghese1 commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1662,6 +1663,51 @@ public <T> T store(final StoreQueryParameters<T> storeQueryParameters) {
+    public boolean isPaused() {
+        if (topologyMetadata.hasNamedTopologies()) {
+            return topologyMetadata.getAllNamedTopologies()
+                .stream().map(NamedTopology::name).allMatch(topologyMetadata::isPaused);
```
Minor: Split across multiple lines for better readability. Eg: https://github.com/apache/kafka/pull/12161/files#diff-b7905c38e362cc37f6989a70b9d0cb028086fbfc79431ddbb9c64870e95482bcL230

cadonna replied:
+1

bvarghese1 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:
```diff
@@ -273,6 +274,12 @@ Collection<Task> allTasks() {
     }
 
+    Collection<Task> notPausedTasks() {
+        return new ArrayList<>(readOnlyActiveTasks).stream().filter(t ->
```
Minor: Split across multiple lines for better readability.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1662,6 +1663,51 @@ public <T> T store(final StoreQueryParameters<T> storeQueryParameters) {
+     */
+    public void pause() {
+        if (topologyMetadata.hasNamedTopologies()) {
+            for (final NamedTopology allNamedTopology : topologyMetadata.getAllNamedTopologies()) {
```
nit: `allNamedTopology` -> `namedTopology` since it is just one.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1662,6 +1663,51 @@ public <T> T store(final StoreQueryParameters<T> storeQueryParameters) {
+    public boolean isPaused() {
+        if (topologyMetadata.hasNamedTopologies()) {
+            return topologyMetadata.getAllNamedTopologies()
+                .stream().map(NamedTopology::name).allMatch(topologyMetadata::isPaused);
```
```suggestion
            return topologyMetadata.getAllNamedTopologies().stream()
                .map(NamedTopology::name)
                .allMatch(topologyMetadata::isPaused);
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1662,6 +1663,51 @@ public <T> T store(final StoreQueryParameters<T> storeQueryParameters) {
+        if (topologyMetadata.hasNamedTopologies()) {
+            return topologyMetadata.getAllNamedTopologies()
+                .stream().map(NamedTopology::name).allMatch(topologyMetadata::isPaused);
+
```
Please remove empty line.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1662,6 +1663,51 @@ public <T> T store(final StoreQueryParameters<T> storeQueryParameters) {
+     */
+    public void resume() {
+        if (topologyMetadata.hasNamedTopologies()) {
+            for (final NamedTopology allNamedTopology : topologyMetadata.getAllNamedTopologies()) {
```
nit: same as above

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -463,7 +463,10 @@ public void restore(final Map<TaskId, Task> tasks) {
                 final TaskId taskId = changelogs.get(partition).stateManager.taskId();
                 try {
                     if (restoreChangelog(changelogs.get(partition))) {
```
Are you sure this method call avoids restoring state stores of paused tasks? Wouldn't `restoringChangelogs()` still return all changelog partitions that are in restoration and not just those that are not paused?
Maybe it is possible to add some verifications to the unit tests to ensure that only non-paused tasks are restored.

jnh5y replied:
I'll look to add a unit test to verify things.

This part may get subtle...  I tried to update things and the ITs showing that one can start topologies in paused state broke.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1662,6 +1663,51 @@ public <T> T store(final StoreQueryParameters<T> storeQueryParameters) {
+     *  Note that pause() can be called before start() in order to start a KafkaStreams instance
+     *  in a manner where the processing is paused as described, but the consumers are started up.
+     */
+    public void pause() {
```
Could you please add unit tests for the new methods?

jnh5y replied:
Yes!  I added some in this commit https://github.com/apache/kafka/pull/12161/commits/e0a80bbf31b79a1eb81bd67dff33ce81933b3e73.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -897,7 +897,8 @@ private void initializeAndRestorePhase() {
         // if there's no active restoring or standby updating it would not try to fetch any data
-        changelogReader.restore(taskManager.tasks());
+        // After KAFKA-13873, we only restore the not paused tasks.
+        changelogReader.restore(taskManager.notPausedTasks());
```
This should also be verified in a unit test with a mock changelog reader.

jnh5y replied:
I think I've covered this with a new test in StreamThreadTest.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutionMetadata.java:
```diff
@@ -35,21 +35,27 @@ public class TaskExecutionMetadata {
+        this.pausedTopologies = pausedTopologies;
     }
 
     public boolean canProcessTask(final Task task, final long now) {
```
I am aware that there are now unit tests for this class, but there are enough different code paths that would justify to add unit tests for this method.

jnh5y replied:
I've added some unit tests for `TaskExecutionMetadata` here: https://github.com/apache/kafka/pull/12161/commits/0a378794226495074b0f38d5e1295ac3ec45238e

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:
```diff
@@ -275,7 +275,7 @@ private void commitSuccessfullyProcessedTasks() {
         int punctuated = 0;
 
-        for (final Task task : tasks.activeTasks()) {
+        for (final Task task : tasks.notPausedTasks()) {
```
Also here unit tests would be great and easily doable.

jnh5y replied:
@cadonna Ok, I tried to add some unit tests for `TaskExecutor` and I think I was needing to mock quite a few things.  Does that seem right?  Or is there an easier way?

jnh5y replied:
Note to self:  Use `notPausedActiveTasks`.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:
```diff
@@ -273,6 +274,12 @@ Collection<Task> allTasks() {
         return readOnlyTasks;
     }
 
+    Collection<Task> notPausedTasks() {
```
Unit tests would be great!

cadonna replied:
Unit tests would still be great!

jnh5y replied:
@cadonna Same deal here, I started to try and test `Tasks` instances and I started down a path of mocking quite a bit of the internal classes.

I think I'm either missing an easier approach or testing these functions directly may require a decent amount of effort.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:
```diff
@@ -257,6 +260,35 @@ public void registerAndBuildNewTopology(final KafkaFutureImpl<Void> future, fina
+     * Pauses a topology by name
+     * @param topologyName Name of the topology to pause
+     */
+    public void pauseTopology(final String topologyName) {
```
These methods could also be unit tested really easily.

jnh5y replied:
Added tests here: https://github.com/apache/kafka/pull/12161/commits/34ec8ac54ac52196bccace8054cd7784719bf409

cadonna reviewed (commented):
Thanks for the PR @jnh5y !

I would like to see a bit more unit testing.

I was wondering whether the pausing of tasks in restoration really works. See my comment in  the `StoreChangelogReader`.

jnh5y committed:
Cleaning up reviewer comments.

jnh5y committed:
Some work on fixing up the StoreChangelogReader.

jnh5y committed:
Updates to the IT.

jnh5y committed:
IntelliJ refactoring to fix complexity.

jnh5y committed:
Adding unit tests for KafkaStreams.

jnh5y committed:
Adding unit test showing that only non-paused, standby tasks are restored.

jnh5y committed:
Found the suppressions for JavaNCSS.

jnh5y committed:
Added TaskExecutionMetadataTest.

jnh5y committed:
Adding TopologyMetadataTest.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -783,6 +822,28 @@ public void shouldThrowOnCleanupWhileRunning() throws InterruptedException {
+            streams.pause();
+            waitForCondition(
+                streams::isPaused,
+                "Streams did not paused");
```
typo: 
```suggestion
                "Streams did not pause");
```

jnh5y replied:
Rolling this into one of my commits.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -783,6 +822,28 @@ public void shouldThrowOnCleanupWhileRunning() throws InterruptedException {
+                streams::isPaused,
+                "Streams did not paused");
+
+            try {
+                streams.cleanUp();
+                fail("Should have thrown IllegalStateException");
+            } catch (final IllegalStateException expected) {
+                assertEquals("Cannot clean up while running.", expected.getMessage());
+            }
```
Wouldn't it be simpler to use `assertThrows()`?

jnh5y replied:
I'll admit to it; I just copy-pasted the test directly above: https://github.com/apache/kafka/blob/34ec8ac54ac52196bccace8054cd7784719bf409/streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java#L816-L821

Is following the existing codebase ok or shall I spend some time to clean it up?

cadonna replied:
I would use `assertThrows()` in the new test. The existing codebase also uses it in some tests like `shouldThrowExceptionSettingUncaughtExceptionHandlerNotInCreateState`.
I would open a new PR to refactor the tests that use `try-fail-catch`.  That is optional.

jnh5y replied:
Fixed in https://github.com/apache/kafka/pull/12161/commits/759d3a8fd9c3a0df1ac541571c7324e67e6c002c

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -415,68 +415,111 @@ public Set<TopicPartition> completedChangelogs() {
 
-        if (!activeRestoringChangelogs().isEmpty() && state == ChangelogReaderState.STANDBY_UPDATING) {
-            throw new IllegalStateException("Should not be in standby updating state if there are still un-completed active changelogs");
+        if (!activeRestoringChangelogs().isEmpty()
+            && state == ChangelogReaderState.STANDBY_UPDATING) {
```
nit: Why this refactoring? The condition fits in one line.

jnh5y replied:
Changing it back.  I am still dialing in my IntelliJ settings.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1761,6 +1761,96 @@ public void shouldUpdateStandbyTask() throws Exception {
+        thread.runOnce();
+
+        assertEquals(10L, store1.approximateNumEntries());
+        assertEquals(4L, store2.approximateNumEntries());
```
Why 4 and not 5?

jnh5y replied:
Good question.  Short answer:  I'm not certain.  I copied the test here: https://github.com/apache/kafka/blob/trunk/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java#L1684-L1762 since it is dealing with updating standby tasks.

My guess is that the method is actually returning an approximated number of entries as suggested.  I tried to verify that, but I got a little lost in the unit test's complexity.

cadonna replied:
Ah, I see! Could you please extract common code (especially the setup code) of those two tests into a method.
We could consider to just check for greater 0.
Additionally to standbys, we should also test if active tasks in restoration are paused.

jnh5y replied:
For the refactoring, I've taken a pass at it here: https://github.com/apache/kafka/pull/12161/commits/a3bd8aec8990e2e72aec1401340c6f457db4f878.

I tend to prefer to add new things consistently with existing code, so I'm hesitating to change the check.  I'm fine either way.

I'll add something for active tasks if that's the way we go.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutionMetadataTest.java:
```diff
@@ -0,0 +1,123 @@
+        // One second after the error, task1 cannot process, task2 can.
+        Assert.assertFalse(metadata.canProcessTask(mockTask1, 1000));
+        Assert.assertTrue(metadata.canProcessTask(mockTask2, 1000));
+
```
Could you add a case where the time since the error is exactly the backoff time (i.e. 5000)?
I would prefer to put the times 1000, 5000, and 10000 into variables with meaningful names instead of adding comments.

jnh5y replied:
Sure.  I took a crack at updating the test; lemme know if you'd prefer different names.

cadonna replied:
Could you please use `CONSTANT_BACKOFF_MS - 1` for the verifications on line 111 and 112? I think it makes the intent clearer.

jnh5y replied:
Fixed in https://github.com/apache/kafka/pull/12161/commits/cfb18f5b61401c28d30a67ac4fbc8d0c1f12ce6f

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -415,68 +415,111 @@ public Set<TopicPartition> completedChangelogs() {
+                try {
+                    // for restoring active and updating standby we may prefer different poll time
+                    // in order to make sure we call the main consumer#poll in time.
+                    // TODO: once we move ChangelogReader to a separate thread this may no longer be a concern
```
I think these comments should be moved before the call to `restoreConsumer.poll()`. BTW, this again confirms that inline comments are a poor way to document code, most of the times.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -415,68 +415,111 @@ public Set<TopicPartition> completedChangelogs() {
-            for (final TopicPartition partition : polledRecords.partitions()) {
-                bufferChangelogRecords(restoringChangelogByPartition(partition), polledRecords.records(partition));
-            }
+                // JNH: Fix this?
```
Please do not forget to remove those comments here and elsewhere.

cadonna reviewed (commented):
Thanks for the updates, @jnh5y!

I did a quick pass. Here my feedback.

jnh5y committed:
Fixes to make the StoreChangelogReader work.

jnh5y committed:
Responding to reviewer comments.

jnh5y committed:
Minor refactoring / renaming.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,445 @@
+    }
+
+    private static void produceToInputTopics(final String topic,
+        final Collection<KeyValue<String, Long>> records) {
```
nit:
```suggestion
    private static void produceToInputTopics(final String topic, final Collection<KeyValue<String, Long>> records) {
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,445 @@
+        final Map<String, Map<Integer, LagInfo>> lagMap = kafkaStreams.allLocalStorePartitionLags();
+        assertNoLag(lagMap);
+
+        // Verify no output somehow?
+        // Is there a better way to show this?
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 0),
+            CoreMatchers.equalTo(Collections.emptyList()));
```
As you write in the comments, this is not straightforward. I think it is not good to just wait for no outputs for two reasons:
- it increases the duration of the test unnecessarily
- it opens the door for flakiness because we only rely on a duration for verification

I have the following idea. You create two identical `KafkaStreams` clients (be careful to specify distinct state directories for each). Both clients read from the same input topics but write to distinct output topics. You do the initial setup verifications for both. Then you pause one of the clients and wait until the other client produced a given amount of records to the output topic. When that happened, you verify if the paused client has not written anything to its output topic. The assumption is that both clients produce with the same rate to their output topics if they are not paused. If one of the two is paused and does not produce any records to the output topic in the same time the other client produces a certain number of records we can assume the pausing works.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,445 @@
+
+    @Test
+    public void shouldPauseAndResumeKafkaStreams() throws Exception {
+        // Create KafkaStreams instance
```
This and the other comments are rather obvious. I would remove them.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,445 @@
+        // Start KafkaStream with paused processing.
+        kafkaStreams.pause();
+        kafkaStreams.start();
+        // Check for rebalancing instead?
```
Why would you like to wait for `REBALANCING` instead of `RUNNING`?

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,445 @@
+
+        // Verify that consumers read new data -- AKA, there is no lag.
+        final Map<String, Map<Integer, LagInfo>> lagMap = kafkaStreams.allLocalStorePartitionLags();
+        assertNoLag(lagMap);
```
Just to be clear, this works as long as the input buffers for polled records are not full. I think it is good enough for this test, though.

jnh5y replied:
Ayup, that's what I understand and it is showing that the active task consumers are still reading (up until their buffers are full).

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,445 @@
+
+
+    @Test
+    public void shouldPauseAndResumeKafkaStreams() throws Exception {
```
Could you please try to extract common code to reusable methods?

jnh5y replied:
Yes.  I've extracted some reusable methods.

cadonna reviewed (commented):
@jnh5y Thank you for the updates!

Here my feedback:

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -479,6 +485,26 @@ public void restore(final Map<TaskId, Task> tasks) {
         }
     }
 
+    private void updateStandbyPartitions(final Map<TaskId, Task> tasks,
```
I think you forgot to also pause the active tasks in restoration.

guozhangwang replied:
I'm wondering if we can make this a more general rule, like:

1) When reader is in Restore_Active state i.e. there are at least one (for sake of simplicity just say we happen to have exactly one) active task which needs restoration, say taskA; then taskA was paused, we should be able to transit to Update_Standby.
2) When reader is in Update_Standby state, and there is one active task say taskA resumed; we should be able to transit to Restore_Active.

I know for now it does not matter since we always pause all tasks with the current APIs, but this is extensible for finer-grained controls in the future.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -479,6 +485,26 @@ public void restore(final Map<TaskId, Task> tasks) {
+                final Task task = tasks.get(taskId);
+                if (manager.taskType() == TaskType.STANDBY) {
+                    if (task != null) {
+                        restoreConsumer.resume(Collections.singleton(partition));
+                    } else {
+                        restoreConsumer.pause(Collections.singleton(partition));
+                    }
```
I think a better way would be to collect the partitions to resume and pause and then call `restoreConsumer.pause()` and `restoreConsumer.restore()` just once with the collections of partitions.

jnh5y replied:
Done in https://github.com/apache/kafka/pull/12161/commits/cb6d6f132a91c3373cbb6556d0b00b1d2bf8eb1b.

jnh5y committed:
Update streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

jnh5y committed:
Merge remote-tracking branch 'origin/trunk' into kafka-13873

jnh5y committed:
Using assertThrows.

jnh5y committed:
Updating TaskExecutorMetadataTest.

jnh5y committed:
Extracted some code in StreamThreadTest.java.

jnh5y committed:
Refactoring StoreChangelogReader.

jnh5y committed:
Fixing a checkstyle issue.

jnh5y committed:
commit 5d25096520503a733b5e520c022d25aaf9d939a7
Author: Jim Hughes <jhughes@confluent.io>
Date:   Thu Jun 2 09:42:13 2022 -0400

    Tests for Tasks.java

commit 36ae36c875bb07cdcc61fc1d88431b71d6beaa87
Author: Jim Hughes <jhughes@confluent.io>
Date:   Thu Jun 2 09:03:44 2022 -0400

    Test for Tasks.java.

commit 094c485063a7961592f4737c45430dc9e21e7161
Author: Jim Hughes <jhughes@confluent.io>
Date:   Wed Jun 1 15:50:58 2022 -0400

    Attempt to create tests for Tasks.java.

jnh5y committed:
Various clean up.

jnh5y committed:
Small clean up.

jnh5y committed:
Adding PR IT and fixing an import.

jnh5y committed:
Test which shows that active tasks still restore.

jnh5y committed:
Pausing active task restoration.

jnh5y committed:
Updated PR IT with multiple client example.  This is hitting SESSION_TIMEOUT_MS_CONFIG.

jnh5y commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,504 @@
+        // Check for rebalancing instead?
+        waitForApplicationState(singletonList(kafkaStreams), State.RUNNING, STARTUP_TIMEOUT);
+        assertTrue(kafkaStreams.isPaused());
+
+        // Create KafkaStreams instance
+        final StreamsBuilder builder2 = new StreamsBuilder();
+        builder2.stream(INPUT_STREAM_1).groupByKey().count().toStream().to(OUTPUT_STREAM_2);
+
+        kafkaStreams2 = new KafkaStreams(builder2.build(props()), props());
+
+        // Start KafkaStream with paused processing.
+        kafkaStreams2.pause();
+        kafkaStreams2.start();
+        // Check for rebalancing instead?
+        //waitForApplicationState(singletonList(kafkaStreams2), State.RUNNING, STARTUP_TIMEOUT);
+        assertTrue(kafkaStreams2.isPaused());
+
+        // Verify no data?
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 0),
+            CoreMatchers.equalTo(Collections.emptyList()));
+
+        // -- Verify that each instance is in rebalancing (after change to pause active task restoration)
+
+        System.out.println("JNH: calling close: " + kafkaStreams.state());
+        // Close the other -- this causes a rebalance
+        kafkaStreams2.close();
+        waitForApplicationState(singletonList(kafkaStreams2), State.NOT_RUNNING, STARTUP_TIMEOUT);
+
+        System.out.println("JNH: called close: " + kafkaStreams.state());
+
+        // Resume paused instance
+        kafkaStreams.resume();
+        System.out.println("JNH: called resume " + kafkaStreams.state());
+        waitForApplicationState(singletonList(kafkaStreams), State.RUNNING, STARTUP_TIMEOUT);
+        System.out.println("JNH: streams is running again " + new Date());
+
+        // Observe all data processed
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 3),
+            CoreMatchers.equalTo(COUNT_OUTPUT_DATA));
+    }
```
@cadonna I added this as a test to show what happens between multiple clients.  I am noticing that this test was taking 45+ seconds to run.

It looks like it is hitting the `ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG` and I'm trying to think through why that may be happening.  Any ideas?

jnh5y committed:
Updating the PauseResume IT so that the multiple instance test runs faster.

jnh5y committed:
Fixing StoreChangelogReaderTest.

jnh5y committed:
Removing most comments.

jnh5y committed:
Added a little getTopicSize function.

jnh5y committed:
Verifying Streams pausing by watching the poll count.
Test clean up.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutionMetadataTest.java:
```diff
@@ -0,0 +1,126 @@
+        final Set<String> pausedTopologies = ConcurrentHashMap.newKeySet();
+
+        final TaskExecutionMetadata metadata = new TaskExecutionMetadata(topologies,
+            pausedTopologies);
```
nit:
```suggestion
        final TaskExecutionMetadata metadata = new TaskExecutionMetadata(topologies, pausedTopologies);
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutionMetadataTest.java:
```diff
@@ -0,0 +1,126 @@
+    @Test
+    public void testCanProcessWithoutNamedTopologies() {
+        final Set<String> topologies = Collections.singleton(UNNAMED_TOPOLOGY);
+        final Set<String> pausedTopologies = ConcurrentHashMap.newKeySet();
```
Out of curiosity, why do you use a `ConcurrentHashMap` here?

jnh5y replied:
I used `ConcurrentHashMap` since that's what `pausedTopologies` is created as elsewhere.

In the unit tests, it can be a `HashSet`, I switched to that.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutionMetadataTest.java:
```diff
@@ -0,0 +1,126 @@
+    public void testNamedTopologiesCanBePausedIndependently() {
+        final Set<String> pausedTopologies = ConcurrentHashMap.newKeySet();
+        final TaskExecutionMetadata metadata = new TaskExecutionMetadata(NAMED_TOPOLOGIES,
+            pausedTopologies);
```
nit:
```suggestion
        final TaskExecutionMetadata metadata = new TaskExecutionMetadata(NAMED_TOPOLOGIES, pausedTopologies);
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -83,67 +137,12 @@
-import static org.junit.Assert.assertThrows;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
-
```
I think you IDE reformatted the imports. The same occurred in `TaskExecutionMetadataTest`. Could you check also in other files? We use more or less the following import order:

all other imports non-static  (sorted alphabetically)
<blank line>
java.* (sorted alphabetically)
javax.* (sorted alphabetically)
<blank line>
static imports  (sorted alphabetically)

However, we are not really consistent across files. Nevertheless we should try to keep that order.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1686,8 +1685,74 @@ public void shouldUpdateStandbyTask() throws Exception {
+        final boolean addActiveTask) throws IOException {
+        final TopicPartition activePartition = new TopicPartition(changelogName1, 2);
         final TopicPartition partition1 = new TopicPartition(changelogName1, 1);
         final TopicPartition partition2 = new TopicPartition(changelogName2, 1);
```
```suggestion
    private void setupThread(final String storeName1,
                             final String storeName2,
                             final String changelogName1,
                             final String changelogName2,
                             final StreamThread thread,
                             final MockConsumer<byte[], byte[]> restoreConsumer,
                             final boolean addActiveTask) throws IOException {
        final TopicPartition activePartition = new TopicPartition(changelogName1, 2);
        final TopicPartition partition1 = new TopicPartition(changelogName1, 1);
        final TopicPartition partition2 = new TopicPartition(changelogName2, 1);
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1686,8 +1685,74 @@ public void shouldUpdateStandbyTask() throws Exception {
+        // let the store1 be restored from 0 to 10; store2 be restored from 5 (checkpointed) to 10
+        for (long i = 0L; i < 10L; i++) {
+            restoreConsumer.addRecord(new ConsumerRecord<>(
+                "stream-thread-test-count-one-changelog",
```
Why do you not pass in the store name or use a class constant? Same below and above.

jnh5y replied:
Extracted a constant.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -479,6 +485,47 @@ public void restore(final Map<TaskId, Task> tasks) {
+    }
+
+    private boolean shouldPause(final Map<TaskId, Task> tasks, final TopicPartition partition, final TaskType taskType) {
+        final ProcessorStateManager manager = changelogs.get(partition).stateManager;
+        final TaskId taskId = manager.taskId();
+        final Task task = tasks.get(taskId);
+        if (manager.taskType() == taskType) {
+            return task == null;
+        }
+        return false;
+    }
```
Why not check the task type earlier? If the task type does not match, you do not need to do anything else. 
```suggestion
        final ProcessorStateManager manager = changelogs.get(partition).stateManager;
        if (manager.taskType() == taskType) {
            final TaskId taskId = manager.taskId();
            final Task task = tasks.get(taskId);
            return task == null;
        }
        return false;
    }
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -479,6 +485,47 @@ public void restore(final Map<TaskId, Task> tasks) {
         }
     }
 
+    private void updateStandbyPartitions(final Map<TaskId, Task> tasks,
```
After your changes that include the restoration of active tasks the name of the method does not reflect what the method does anymore.  Please rename the method to something more appropriate.

cadonna replied:
Do not forget to rename this method to something more meaningful. 
Proposal: `pauseResumePartitions()`

cadonna commented on checkstyle/suppressions.xml:
```diff
@@ -225,7 +225,7 @@
 
     <suppress checks="JavaNCSS"
-              files="(EosV2UpgradeIntegrationTest|KStreamKStreamJoinTest|TaskManagerTest).java"/>
+              files="(EosV2UpgradeIntegrationTest|KStreamKStreamJoinTest|StreamThreadTest|TaskManagerTest).java"/>
```
In general, I think we should try to avoid adding suppressions. But I also see that `StreamThreadTest` would need quite some love at the moment which is not the intent of this PR.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,380 @@
+        produceToInputTopics(INPUT_STREAM_1, STANDARD_INPUT_DATA);
+        assertNoLag(kafkaStreams);
+
+        waitUntilStreamsHasPolled(kafkaStreams, 2);
```
I like your approach!

jnh5y replied:
Thanks!  It was either going to be a good idea or prove too hacky!  Glad you like it!

cadonna replied:
It is hacky, but still a good idea 🙂

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,380 @@
+        assertTrue(kafkaStreams.isPaused());
+
+        produceToInputTopics(INPUT_STREAM_1, STANDARD_INPUT_DATA);
+        assertNoLag(kafkaStreams);
```
I just realized that this method computes the lag of the store regarding the changelog partition and not of the input partitions. Was this intentional? I think the lag of the store regarding the changelog topic will always be zero in this case because in normal mode the Streams client writes to the changelog topic and so it is always up-to-date.
I thought that this method verified that the main consumer does not have any lag regarding the input partitions which would tell us that the main consumer polled data although it was paused which is the expected behavior. This method with `waitUntilStreamsHasPolled(kafkaStreams, 2);` and `assertTopicSize(OUTPUT_STREAM_1, <same size as before the call to pause>);` would tell us that although the Streams clients polled data and went through the poll loop, no data was produced to the output topic. For that you should also set the cache size to 0 with `STATESTORE_CACHE_MAX_BYTES_CONFIG`.

jnh5y replied:
First, good catch!  It wasn't intentional...  I thought that I had found the right calls to show no lag for consumers...

I tried to switch over to verifying that there was no consumer lag, and I don't think we can sensibly do so.  

(If I understand correctly) The consumer offsets are only committed after the input records have been processed.  Since there is no processing, the consumer offsets will appear to be behind.  (I verified this locally.)

My conclusion is that the method isn't checking anything useful, so I am planning on removing it.

cadonna replied:
I agree that you should remove `assertNoLag()`.

jnh5y replied:
Removed!

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,380 @@
+    }
+
+    @Test
+    public void pauseResumehouldWorkAcrossInstances() throws Exception {
```
Typo: ` pauseResumehouldWorkAcrossInstances` -> ` pauseResumeShouldWorkAcrossInstances`

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,380 @@
+
+        awaitOutput(OUTPUT_STREAM_1, 3, COUNT_OUTPUT_DATA);
+    }
+
```
Could you please also add a test that verifies if active tasks in restoration and standbys are paused? Something like you start two Streams clients with 1 standby. If you have one partition one client should get the active stateful task and the other should get the standby task. The clients process some data and write some data into their states. Then shutdown the Streams clients and wipe out the local state. Finally, start both clients paused and verify if the lag of the local stores stays constant and greater than zero for a couple of poll loop iterations.

jnh5y replied:
I like the idea, but based on https://github.com/apache/kafka/pull/12161/files#r893999817 I'm not sure if we can sensibly check something useful.

It seems like the offsets (at least for the main consumer) will not be committed, so they will be definitely be at 0 since processing is paused.

For the restore consumer, does it have a groupId?  Would we be able to see anything about its state?

cadonna replied:
The comment https://github.com/apache/kafka/pull/12161/files#r893999817 is unrelated here. In my proposal, you would assert that the lag between local state store and changelog topic does not decrease. You can measure that in a similar way as you already do in `assertNoLag()`.
The restore consumer does not have a group since the partitions are manually assigned to the consumer. If there is no group there are no committed offsets.

jnh5y replied:
I'm still a little light on the details.  I'll give it a try here in a little bit.

jnh5y replied:
Ok, I added a test named `pausedTopologyShouldNotRestoreStateStores` that hopefully will cover things.

cadonna reviewed (commented):
@jnh5y Thank you for the updates!

Here my feedback!

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,380 @@
+        kafkaStreams2 = buildKafkaStreams(OUTPUT_STREAM_2);
+        kafkaStreams2.pause();
+        kafkaStreams2.start();
+        assertTrue(kafkaStreams2.isPaused());
```
Do you also want to wait for `RUNNING` for the second `KafkaStreams` instance?

jnh5y replied:
Yes, I think that's fair.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,380 @@
+        properties.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.Long().getClass());
+        properties.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 1000L);
+        properties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        properties.put(StreamsConfig.TOPOLOGY_OPTIMIZATION_CONFIG, StreamsConfig.OPTIMIZE);
```
I think you can remove this config. It is not relevant for what this test tests.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,380 @@
+    }
+
+    @Test
+    public void pauseResumehouldWorkAcrossInstances() throws Exception {
```
According to the builds, this test seems flaky. I think the reason is that the cache (in Streams, not the RocksDB cache) of the state stores is not set to zero. When the cache is not set to zero, the number of results that are send downstreams is not deterministic, because some intermediate results might be sent downstream and some not. I tried to set the cache to zero and was able to run 100+ times the test in a row without failure whereas with the cache > 0, the test failed much earlier.
Note that by changing the cache size also the results that you verify change.

jnh5y replied:
Thanks.  I was having trouble shifting through the build info.  I've updated the test to set the state store size to 0.  

I'll watch the CI jobs to see if they are green after my last push.  (They seemed to be ok as I was working today.)

cadonna replied:
I think you fixed it! 🎉

jnh5y committed:
Update streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutionMetadataTest.java

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

jnh5y committed:
Update streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutionMetadataTest.java

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

jnh5y committed:
Update streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

jnh5y committed:
Updating import order and sorting out other issues.

jnh5y committed:
Responding to more reviewer comments.

jnh5y committed:
Small test changes before adding new test.

jnh5y committed:
Partial unit test?

jnh5y committed:
Added new unit test.

jnh5y committed:
Removing assertNoLag.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -335,7 +333,39 @@ public void pauseResumehouldWorkAcrossInstances() throws Exception {
+
+        waitForApplicationState(Arrays.asList(kafkaStreams, kafkaStreams2), State.REBALANCING, STARTUP_TIMEOUT);
+
+        assertTrue(kafkaStreams.allLocalStorePartitionLags().isEmpty());
+        assertTrue(kafkaStreams2.allLocalStorePartitionLags().isEmpty());
```
Why are you verifying for emptiness? I would expect that there are entries for the state stores with a lag greater than 0.

cadonna replied:
I played a bit around with the test and indeed if you add a `Thread.sleep(2000)` before these asserts, the test fails because the returned map is not empty. That means, the assignment was not finished before the asserts were called.

cadonna replied:
You could do something like:
```
        waitForApplicationState(Arrays.asList(kafkaStreams), State.REBALANCING, STARTUP_TIMEOUT);
        waitForCondition(
            () -> !kafkaStreams.allLocalStorePartitionLags().isEmpty(),
            "Lags for local store partitions were not found within the timeout!");
        waitUntilStreamsHasPolled(kafkaStreams, 2);
        final long stateStoreLag1 = kafkaStreams.allLocalStorePartitionLags().get("test-store").get(0).offsetLag();
        waitUntilStreamsHasPolled(kafkaStreams, 2);
        final long stateStoreLag2 = kafkaStreams.allLocalStorePartitionLags().get("test-store").get(0).offsetLag();
        assertTrue(stateStoreLag1 > 0);
        assertEquals(stateStoreLag1, stateStoreLag2);
```
This code just considers one Streams client. You need to add `Materialized.as("test-store")` to the call to `count()` in your topology.
As soon as you activated the standbys, you need to do the same for the second Streams client.

jnh5y replied:
Thank you!  I've added these changes.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -335,7 +333,39 @@ public void pauseResumehouldWorkAcrossInstances() throws Exception {
+    public void pausedTopologyShouldNotRestoreStateStores() throws Exception {
+        produceToInputTopics(INPUT_STREAM_1, STANDARD_INPUT_DATA);
+
+        kafkaStreams = buildKafkaStreams(OUTPUT_STREAM_1);
+        kafkaStreams2 = buildKafkaStreams(OUTPUT_STREAM_1);
```
If you do not use standby tasks, there is no reason to use two Kafka Streams clients. I would propose to use one standby only for this test. For that you need to set `num.standby.replicas` to 1. That has the effect that one client gets the active store assigned and the other gets the standby store assigned.

jnh5y replied:
My mistake; I've updated the test.

cadonna reviewed (commented):
@jnh5y Thanks you for the updates!

Here my feedback.

jnh5y committed:
Responding to reviewer comments.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java:
```diff
@@ -0,0 +1,422 @@
+
+        waitForApplicationState(Arrays.asList(kafkaStreams, kafkaStreams2), State.REBALANCING, STARTUP_TIMEOUT);
+
+        assertStreamsLagStaysConstant(kafkaStreams);
+        assertStreamsLagStaysConstant(kafkaStreams2);
```
nit:  `assertStreamsLagStaysConstant()` -> `assertStreamsLocalStoreLagStaysConstant()`

jnh5y replied:
Changed it!

cadonna reviewed (approved):
@jnh5y Thank you for the updates!

LGTM!

Had just one nit.

Thank you for your patience!

jnh5y committed:
Responding to reviewer comments.

jnh5y commented:
> 



> @jnh5y Thank you for the updates!
> 
> LGTM!
> 
> Had just one nit.
> 
> Thank you for your patience!

@cadonna Thank you for pushing me and helping me learn more about streams!
